servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java

```diff
@@ -59,17 +116,75 @@ public static synchronized void initializeFileSystem(boolean permissionCheckEnab
    * @param groups in which user belongs to
    * @param action requested {@link FileSystemAction} by user
    * @param inodes all the inodes retrieved by traversing the path
+   * @param selfIndex the index of the last inode in the inodes list
+   * @param parentOrAncestorIndex the index of the parent or ancestor inode in the inodes list
+   * @param isCheckOwner indicates whether to check that the user is the owner of the last inode
    * @throws IOException if permission checking fails
    */
-  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
-      List<Inode> inodes) throws IOException {
+  private static void checkPermission(String user, List<String> groups, FileSystemAction action,
+      List<Inode> inodes, int selfIndex, int parentOrAncestorIndex, boolean isCheckOwner)
+      throws IOException {
+    int size = inodes.size();
+    assert size > 0;
+
     if (!sPermissionCheckEnabled) {
       return;
     }
 
-    if (sFileSystemOwner.equals(user.getName()) || groups.contains(sFileSystemSuperGroup)) {
+    if (sFileSystemOwner.equals(user) || groups.contains(sFileSystemSuperGroup)) {
       return;
     }
-    // TODO(dong): Implement TACHYON-1416
+
+    // traverses parent path to ensure inodes in it are all executable
+    for (int i = 0; i < size - 1; i++) {
+      check(user, groups, inodes.get(i), FileSystemAction.EXECUTE);
+    }
+
+    // check self permission if needed
+    if (selfIndex > -1) {
+      check(user, groups, inodes.get(selfIndex), action);
+    }
+
+    // check parent or ancestor permission if needed
+    if (parentOrAncestorIndex > -1) {
+      check(user, groups, inodes.get(parentOrAncestorIndex), action);
+    }
+
+    // check owner if needed
+    if (isCheckOwner) {
+      if (user.equals(inodes.get(size - 1).getUserName())) {
+        return;
+      }
+      throw new IOException(ExceptionMessage.PERMISSION_DENIED.getMessage(toExceptionMessage()));
+    }
+  }
+
+  private static void check(String user, List<String> groups, Inode inode, FileSystemAction action)
+      throws IOException {
+    if (inode == null) {
+      return;
+    }
+
+    short permission = inode.getPermission();
+    if (user.equals(inode.getUserName())) {
+      if (FileSystemPermission.getUserAction(permission).imply(action)) {
+        return;
+      }
+    } else if (groups.contains(inode.getGroupName())) {
```

Comment:
seeing as your current unit tests don't catch this problem, you should add a test for this